staging/src/k8s.io/apiserver/pkg/storage/etcd3/size_monitor.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/etcd3/metrics"
+	"k8s.io/klog/v2"
+)
+
+const dbMetricsMonitorJitter = 0.5
+
+var (
+	// dbMetricsMonitorsMu guards access to dbMetricsMonitors map
+	dbMetricsMonitorsMu sync.Mutex
+	dbMetricsMonitors   map[string]struct{}
+)
+
+// StartDBSizeMonitor starts a loop to monitor etcd database size and update the
+// corresponding metric etcd_db_total_size_in_bytes for each etcd server
+// endpoint. The monitors are stopped when the supplied context is done.
+func StartDBSizeMonitor(ctx context.Context, client *clientv3.Client, interval time.Duration) {
+	if interval == 0 {
+		return
+	}
+	dbMetricsMonitorsMu.Lock()
+	defer dbMetricsMonitorsMu.Unlock()
+
+	for _, ep := range client.Endpoints() {
```

Comment:
just realized that endpoints are always added never deleted,
 if an endpoint changes and no longers exist it will be still be polled, right?

Reply:
It looks that way, but I think that was always the case right

Reply:
I was going to raise a tracking issue, but my read of the code is that endpoints can only be removed by restarting the API server process. As far as I can tell they're only added by the CLI flag linked below.

https://github.com/kubernetes/kubernetes/blob/1a916f2/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L156

Reply:
yeah, it seems so, I was just curious, thanks for doing the investigation